server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4317,17 +4317,15 @@ public void resetEngine() {
         assert waitForEngineOrClosedShardListeners.isDone();
         try {
             synchronized (engineMutex) {
-                final var currentEngine = getEngine();
-                currentEngine.prepareForEngineReset();
-                var engineConfig = newEngineConfig(replicationTracker);
                 verifyNotClosed();
-                IOUtils.close(currentEngine);
-                var newEngine = createEngine(engineConfig);
-                currentEngineReference.set(newEngine);
+                getEngine().prepareForEngineReset();
+                var newEngine = createEngine(newEngineConfig(replicationTracker));
+                IOUtils.close(currentEngineReference.getAndSet(newEngine));
                 onNewEngine(newEngine);
             }
             onSettingsChanged();
         } catch (Exception e) {
+            assert false : e;
```

Comment:
I'm not sure we should assert here, it's a possible scenario.

For example, in the stress test, when we enable the object store failures (currently commented out), it can fail the flush of the hollow commit, and the exception (e.g., Corruption exception) will come here and fail the shard. And we'd like to be able to test that in the stress test. So I'd suggest we remove this assertion.